metron/core/requests_launcher.py

```diff
@@ -1,49 +1,107 @@
+import asyncio
 from typing import Any, List
 
 from ray.util import ActorPool
 
-from metron.core.llm_clients.base_llm_client import BaseLLMClient
+from metron.core.llm_clients import construct_clients
 from metron.core.request_config import RequestConfig
+from metron.core.requests_manager import AsyncRequestsManager
 
 
 class RequestsLauncher:
     """Launch requests from LLMClients to their respective LLM APIs."""
 
-    def __init__(self, llm_clients: List[BaseLLMClient]):
-        self.llm_client_pool = ActorPool(llm_clients)
+    def __init__(
+        self,
+        model: str,
+        llm_api: str,
+        num_ray_clients: int,
+        num_concurrent_requests_per_client: int,
+    ):
+        llm_clients = construct_clients(
+            model_name=model,
+            llm_api=llm_api,
+            num_clients=num_ray_clients,
+            use_ray=False,
+        )
+        self.actors = []
+        for client_id, client in enumerate(llm_clients):
+            self.actors.append(
+                AsyncRequestsManager.remote(
+                    client_id,
+                    client,
+                    max_concurrent_requests=num_concurrent_requests_per_client,
+                )
+            )
+        self.llm_client_pool = ActorPool(self.actors)
+
+    async def start(self) -> None:
+        """Starts the tasks on each actor to handle requests.
+
+        Returns:
+            None
+
+        """
+        for actor in self.actors:
+            await actor.start_tasks.remote()
 
-    def launch_requests(self, request_config: RequestConfig) -> None:
+    async def launch_requests(self, request_config: RequestConfig) -> None:
         """Launch requests to the LLM API.
 
         Args:
             request_config: The configuration for the request.
 
         """
-        if self.llm_client_pool.has_free():
-            self.llm_client_pool.submit(
-                lambda client, _request_config: client.send_llm_request.remote(
-                    _request_config
-                ),
-                request_config,
-            )
+        self.llm_client_pool.submit(
+            lambda actor, _request_config: actor.launch_requests.remote(
+                _request_config
+            ),
+            request_config,
+        )
+
+    async def is_free(self) -> bool:
+        """Check if the pool of actors is free.
+
+        Returns:
+            True if the pool of actors is free, False otherwise.
 
-    def get_next_ready(self, block: bool = False) -> List[Any]:
-        """Return results that are ready from completed requests.
+        """
+        return self.llm_client_pool.has_free()
+
+    async def free_pool(self, block: bool = False) -> None:
```

Comment:
i couldn't follow this exactly. Can you please describe it briefly?

Reply:
So the logic for pool works as follows:
1. New requests are submitted to pool of n ray actors.
2. Each ray actor takes the request and adds it to queue to be picked up by one of its tasks. 
    * Once ray actor is able to add request to queue, it becomes free to take new requests.
    * But if queue is already full, ray actor will remain busy until it is able to add new request to queue.
3. So `is_free()` is to check whether any ray actor is free to take new requests.
4. And `free_pool()` just frees the pool with ray actors who have been able to launch request successfully, so that new requests can be launched.